pkg/migration/plan_generator.go

```diff
@@ -0,0 +1,501 @@
+	errCompositeOutput         = "failed to output migrated composite"
+	errClaimOutput             = "failed to output migrated claim"
+	errClaimsEdit              = "failed to edit claims"
+	errPlanGeneration          = "failed to generate the migration plan"
+	errPause                   = "failed to store a paused manifest"
+)
+
+type step int
+
+const (
+	stepPauseManaged step = iota
+	stepPauseComposites
+	stepCreateNewManaged
+	stepNewCompositions
+	stepEditComposites
+	stepEditClaims
+	stepDeletionPolicyOrphan
+	stepDeleteOldManaged
+	stepStartManaged
+	stepStartComposites
+)
+
+// PlanGenerator generates a migration.Plan reading the manifests available
+// from `source`, converting managed resources and compositions using the
+// available `migration.Converter`s registered in the `registry` and
+// writing the output manifests to the specified `target`.
+type PlanGenerator struct {
+	source   Source
+	target   Target
+	registry Registry
+	// Plan is the migration.Plan whose steps are expected
+	// to complete a migration when they're executed in order.
+	Plan Plan
+}
+
+// NewPlanGenerator constructs a new PlanGenerator using the specified
+// Source and Target and the default converter Registry.
+func NewPlanGenerator(source Source, target Target) PlanGenerator {
+	return PlanGenerator{
+		source:   source,
+		target:   target,
+		registry: registry,
+	}
+}
+
+// GeneratePlan generates a migration plan for the manifests available from
+// the configured Source and writing them to the configured Target using the
+// configured converter Registry. The generated Plan is available in the
+// PlanGenerator.Plan variable if the generation is successful
+// (i.e., no errors are reported).
+func (pg *PlanGenerator) GeneratePlan() error {
+	pg.buildPlan()
+	return errors.Wrap(pg.convert(), errPlanGeneration)
+}
+
+func (pg *PlanGenerator) convert() error { //nolint: gocyclo
+	convertedMR := make(map[corev1.ObjectReference][]UnstructuredWithMetadata)
+	convertedComposition := make(map[string]string)
+	var composites []UnstructuredWithMetadata
+	var claims []UnstructuredWithMetadata
+	for hasNext, err := pg.source.HasNext(); ; hasNext, err = pg.source.HasNext() {
+		if err != nil {
+			return errors.Wrap(err, errSourceHasNext)
+		}
+		if !hasNext {
+			break
+		}
+		o, err := pg.source.Next()
+		if err != nil {
+			return errors.Wrap(err, errSourceNext)
+		}
+		switch gvk := o.Object.GroupVersionKind(); gvk {
+		case xpv1.CompositionGroupVersionKind:
+			target, converted, err := pg.convertComposition(o)
+			if err != nil {
+				return errors.Wrap(err, errCompositionMigrate)
+			}
+			if converted {
+				migratedName := fmt.Sprintf("%s-migrated", o.Object.GetName())
+				convertedComposition[o.Object.GetName()] = migratedName
+				target.Object.SetName(migratedName)
+				if err := pg.stepNewComposition(target); err != nil {
+					return errors.Wrap(err, errCompositionMigrate)
+				}
+			}
+		default:
+			if o.Metadata.IsComposite {
```

Comment:
nit: handling these `IsComposite` and `IsClaim` as cases in `switch` could make it easier to read.